File: C++/MEDIUM/1653. Minimum Deletions to Make String Balanced.cpp
class Solution {
public:
    int minimumDeletions(string s) {
        vector<int> shouldDb (s.length(), 0);
        vector<int> shouldDa (s.length(), 0);

        for (int a = 0; a < s.length(); a++)
        {
            shouldDb[a] = (a ? shouldDb[a - 1] : 0) + (s[a] == 'b');
        }
        for (int b = s.length() - 1; b >= 0; b--)
        {
            shouldDa[b] = b != s.length() - 1 ? shouldDa[b + 1] + (s[b + 1] == 'a') : 0;
        }

        // if (shouldDb[s.length() - 1] == s.length()) return 0;

        int deleteIndex = 0;
        int answer = 100001;
        for (int c = 0; c < s.length(); c++)
        {
            // cout << shouldDa[c] << ' ' << shouldDb[c] << '\n';
            if (shouldDa[c] + shouldDb[c] < answer) 
            {
                answer = shouldDa[c] + shouldDb[c];
                deleteIndex = c;
            }
        }
        // cout << deleteIndex << '\n';
        if (!deleteIndex && shouldDb[0] == 1)
        {
            answer--;
        }
        return answer;
    }
};

File: 1653-minimum-deletions-to-make-string-balanced/1653-minimum-deletions-to-make-string-balanced.java
class Solution {
    public int minimumDeletions(String s) {
        int n = s.length();
        int[] aLeft = new int[n];
        int[] bRight = new int[n];
        for (int i = 0; i < n; i++)
        {
            aLeft[i] = i == 0 ? 0 : aLeft[i - 1]; 
            if (s.charAt(i) == 'b')
            {
                aLeft[i]++;
            }
        }
        int answer = aLeft[n - 1];
        for (int i = n - 1; i >= 0; i--)
        {
            bRight[i] = i == n - 1 ? 0 : bRight[i + 1]; 
            if (s.charAt(i) == 'a')
            {
                bRight[i]++;
            }
            if (i != n - 1) answer = Math.min(aLeft[i] + bRight[i + 1], answer);
            if (i != 0) answer = Math.min(answer, aLeft[i - 1] + bRight[i]);
        }
        answer = Math.min(answer, bRight[0]);
        return answer;
    }
}